### ISSUE TITLE
Combine demos to one binary

### ISSUE THREAD

**Issue by bluetech** [xkbcommon/libxkbcommon#151]
I think it will be nice to combine all of the demo programs we have to one single `xkbcommon` demo binary. It will be able to:

- Show the version it was compiled with, include paths, build-time constants, enabled features etc.
- Interactive evdev demo (on Linux)
- Interactive X11 demo (when xkbcommon-x11 is enabled)
- Interactive Wayland demo (when wayland is enabled)
- Print RMLVO to KCCGST conversion
- Print keymap from RMLVO
- Print (full compiled) keymap from file
- "How to type" functionality
- Benchmarks (compose, key proc, rules, rulescomp)

It should use a `./xkbcommon <subcommand>` UI, with consistent option processing.

All of this already exists but is spread around a dozen different binaries.

**Comment by whot** [comment]
would probably be useful to move this to `tools/` as well to keep the unit-tests separate from tools

**Comment by bluetech** [comment]
@whot Do you think it's a good idea? The `xinput` tool was an inspiration for this :)

**Comment by whot** [comment]
a common prefix is a good idea, makes it easier to find the tools with `./builddir/xkbcommon-<tab><tab>` when debugging. A single tool depends on the use-case you envision. If you expect users to execute those tools (quick check of `meson.build` shows we don't install them) then having a single tool is primarily interesting because of the namespace pollution otherwise.

for `xinput` I have a single binary, for libinput I picked the git-style approach of a parent tool that `execve`s into the subtools in `/usr/libexec`. Biggest advantage here is that the tools can be in e.g. python as well. In the builddir, they're still all `libinput-foo-bar`. I think that's the best approach, esp. if you don't quite know what the future will bring.

If you don't really expect those tools to ever be used outside the git build dir, then namespacing is enough, the effort to have a single tool just gets in the way of the debugger.

**Comment by whot** [comment]
I had a look at this today and started thinking - what tools do we actually need and what for? Here are a few thoughts.

We have four groups of tools atm:
- `interactive-evdev`, `interactive-wayland`, `interactive-x11`. Those *could* be merged into one tool with a `--backend=evdev` option or somesuch, together with making the output fully consistent across all. 
- `registry-list`: I think the output format needs to be improved to make it end-user compatible
- `how-to-type`: standalone tool
- `print-compiled-keymap`, `rmlvo-to-kccgst`, `rmlvo-to-keymap`: these are all more or less the same tool anyway and could easily be combined into one tool. This is #159 now.

So for a single tool, I think the following CLI would be good:
```
$> xkbtool compile-keymap  # handles any compilation
$> xkbtool list # handles the registry
$> xkbtool debug {evdev|wayland|x11} # the interactive tools
$> xkbtool query how-to-type
```
The `compile-keymap` functionality we'll need to eventually install because it needs to be the debugging tool for #108. The `list` functionality is needed because with #146 the include paths matters. The interactive debugging tools should be named as such to prevent any expectation of a stable interface.

The slightly odd-one-out is `how-to-type`, not sure how to categorise that one but grouping it behind `xkbtool query` may not be the worst idea.

Whether it's `xkbcommontool` or `xkbcommon` or `xkbtool` or `blue-purple-banana-fairies` I'll leave up to you to decide ;)

**Comment by bluetech** [comment]
This plan sounds great to me, although I do prefer `interactive` over `debug` because `debug` is too general and not very indicative on what it does (although `interactive` is not a great name either). But if from your experience people really expect stability from such interactive programs, I trust your judgement on that :)

> Whether it's xkbcommontool or xkbcommon or xkbtool or blue-purple-banana-fairies I'll leave up to you to decide ;)

`xkbtool` is too similar to [xdotool](https://www.semicomplete.com/projects/xdotool/) I think. I think `xkbcommon` is good, or even just `xkb` if we're really presumptuous :)

**Comment by whot** [comment]
hehe, we can leave `interactive-debug` which has the best of both worlds and the extra typing can be avoided by shipping completion files. I think just `xkb` is risky because of namespace pollution and searching for it is going to be terrible. We'll also have to then deal with any XKB-related bug report because if we ship `xkb`, clearly any mention of XKB online will be our bug to triage.

I'll use `xkbcommon` for now though I'd prefer it to have `tool` or `cli` or `cmd` or something in it's name to make it more obvious for what it is. How do you like `xkbcmd` or `xkbcli`? I dont' think it's too presumptuous, I reckon it'll take a while before someone  implements a competing version of XKB and they can just use the beloved `ng-` prefix then ;)

**Comment by bluetech** [comment]
`xkbcli` sounds good to me!

**Comment by whot** [comment]
a note on the naming - I started with tools like `xkbcommon interactive-debug wayland` but in hindsight this is too much indirection. There's no real benefit of having that double-nesting with `interactive-debug` in between so it's better to just have `xkbcommon interactive-wayland` instead. Or `xkbcli` as already implemented in #160.

Same goes for `how-to-type`, hiding this behind an extra `query` step is not likely to add much.
